// fynixui/dist/hooks/nixFor.js
import { Fragment } from "../runtime";
export function For(props) {
    let items = [];
    if (props.each && typeof props.each === "object" && "_isNixState" in props.each) {
        items = props.each.value;
    }
    else if (Array.isArray(props.each)) {
        items = props.each;
    }
    let renderer;
    if (typeof props.children === "function") {
        renderer = props.children;
    }
    else if (Array.isArray(props.children)) {
        const firstChild = props.children[0];
        if (typeof firstChild === "function") {
            renderer = firstChild;
        }
    }
    if (!renderer) {
        if (items.length > 0) {
            console.warn("[Fynix] <For> expects a function as its child. Received:", typeof props.children);
        }
        return { type: Fragment, props: { children: [] }, key: null };
    }
    const mapped = items.map((item, index) => {
        try {
            return renderer(item, index);
        }
        catch (error) {
            console.error(`[Fynix] Error rendering item at index ${index}:`, error);
            return {
                type: "div",
                props: {
                    children: ["Error rendering item"],
                    style: "color: red;"
                },
                key: index
            };
        }
    });
    return { type: Fragment, props: { children: mapped }, key: null };
}
