// fynixui/dist/hooks/nixLazyAsync.js
import { activeContext } from "../context/context.js";
import { nixState } from "./nixState.js";
export function nixLazyAsync(importFn, options = {}) {
    const { retry = 0 } = options;
    const cache = {
        status: "pending",
        component: null,
        error: null,
        promise: null,
        retriesLeft: retry,
    };
    let canceled = false;
    const loadModule = () => {
        cache.promise = importFn()
            .then((module) => {
            if (!canceled) {
                cache.status = "success";
                cache.component = module.default || module;
            }
            return cache.component;
        })
            .catch((err) => {
            if (!canceled) {
                if (cache.retriesLeft > 0) {
                    cache.retriesLeft--;
                    return loadModule();
                }
                cache.status = "error";
                cache.error = err;
            }
            return Promise.reject(cache.error);
        });
        return cache.promise;
    };
    loadModule();
    return function LazyWrapper(props) {
        const ctx = activeContext;
        if (!ctx)
            throw new Error("nixLazyAsync() called outside component");
        if (cache.status === "pending") {
            throw cache.promise;
        }
        if (cache.status === "error") {
            throw cache.error;
        }
        if (!cache.component) {
            throw new Error("nixLazyAsync: component not loaded");
        }
        return cache.component(props);
    };
}
export function Suspense({ fallback, children, }) {
    const loading = nixState(false);
    const error = nixState(null);
    try {
        return children();
    }
    catch (promise) {
        if (promise instanceof Promise) {
            loading.value = true;
            promise
                .then(() => (loading.value = false))
                .catch((err) => {
                error.value = err;
                loading.value = false;
            });
            return fallback;
        }
        throw promise;
    }
}
