// fynixui/dist/hooks/nixState.js
import { activeContext } from "../context/context";
export function nixState(initial) {
    const ctx = activeContext;
    if (!ctx)
        throw new Error("nixState() called outside component");
    const idx = ctx.hookIndex++;
    if (!ctx.hooks[idx]) {
        let value = initial;
        const subscribers = new Set();
        let isDestroyed = false;
        if (initial !== null && typeof initial === "object") {
            if ("__proto__" in initial ||
                "constructor" in initial ||
                "prototype" in initial) {
                if (Array.isArray(initial)) {
                    value = [...initial];
                }
                else {
                    value = { ...initial };
                    delete value.__proto__;
                    delete value.constructor;
                    delete value.prototype;
                }
            }
        }
        const s = {
            get value() {
                if (isDestroyed) {
                    return value;
                }
                if (activeContext && !isDestroyed) {
                    activeContext._accessedStates.add(s);
                }
                return value;
            },
            set value(newVal) {
                if (isDestroyed) {
                    return;
                }
                if (newVal === value)
                    return;
                if (newVal !== null && typeof newVal === "object") {
                    if ("__proto__" in newVal ||
                        "constructor" in newVal ||
                        "prototype" in newVal) {
                        if (Array.isArray(newVal)) {
                            newVal = [...newVal];
                        }
                        else {
                            newVal = { ...newVal };
                            delete newVal.__proto__;
                            delete newVal.constructor;
                            delete newVal.prototype;
                        }
                    }
                }
                value = newVal;
                const subsArray = Array.from(subscribers);
                subsArray.forEach((fn) => {
                    try {
                        fn(newVal);
                    }
                    catch (err) {
                        console.error("[nixState] Subscriber error:", err);
                        subscribers.delete(fn);
                    }
                });
            },
            subscribe(fn) {
                if (typeof fn !== "function") {
                    console.error("[nixState] subscribe() requires a function");
                    return () => { };
                }
                if (isDestroyed) {
                    console.warn("[nixState] Cannot subscribe to destroyed state");
                    return () => { };
                }
                const MAX_SUBSCRIBERS = 1000;
                if (subscribers.size >= MAX_SUBSCRIBERS) {
                    console.error("[nixState] Maximum subscriber limit reached");
                    return () => { };
                }
                subscribers.add(fn);
                return () => {
                    subscribers.delete(fn);
                };
            },
            cleanup() {
                if (isDestroyed)
                    return;
                isDestroyed = true;
                subscribers.clear();
                if (value !== null && typeof value === "object") {
                    if (Array.isArray(value)) {
                        value = [];
                    }
                    else {
                        value = null;
                    }
                }
                console.log("[nixState] State cleaned up");
            },
            getSubscriberCount() {
                return subscribers.size;
            },
            isDestroyed() {
                return isDestroyed;
            },
            asReadOnly() {
                return {
                    get value() {
                        return s.value;
                    },
                    subscribe: s.subscribe.bind(s),
                    _isNixState: true,
                    _isReadOnly: true,
                };
            },
            _isNixState: true,
        };
        ctx.hooks[idx] = s;
        if (ctx.stateCleanups) {
            ctx.stateCleanups.push(() => s.cleanup());
        }
    }
    return ctx.hooks[idx];
}
