// fynixui/dist/plugins/vite-plugin-res.js
import { transform } from "esbuild";
import { normalizePath } from "vite";
export default function fynixPlugin(options = {}) {
    const { jsxFactory = "Fynix", jsxFragment = "Fynix.Fragment", include = [".ts", ".js", ".jsx", ".tsx", ".fnx"], exclude = ["node_modules"], sourcemap = true, esbuildOptions = {}, } = options;
    let viteServer = null;
    return {
        name: "vite-plugin-fynix",
        enforce: "pre",
        configureServer(server) {
            viteServer = server;
        },
        async transform(code, id) {
            const normalizedId = normalizePath(id);
            const shouldExclude = exclude.some((pattern) => normalizedId.includes(pattern));
            if (shouldExclude)
                return null;
            const shouldInclude = include.some((ext) => normalizedId.endsWith(ext));
            if (!shouldInclude)
                return null;
            const ctx = this;
            if (typeof ctx.addWatchFile === "function") {
                ctx.addWatchFile(id);
            }
            try {
                let loader = "tsx";
                if (normalizedId.endsWith(".ts") || normalizedId.endsWith(".fnx")) {
                    loader = "tsx";
                }
                else if (normalizedId.endsWith(".tsx")) {
                    loader = "tsx";
                }
                else if (normalizedId.endsWith(".jsx")) {
                    loader = "jsx";
                }
                else if (normalizedId.endsWith(".js")) {
                    loader = "jsx";
                }
                const result = await transform(code, {
                    loader,
                    jsxFactory,
                    jsxFragment,
                    sourcemap,
                    sourcefile: id,
                    target: "esnext",
                    format: "esm",
                    ...esbuildOptions,
                });
                return {
                    code: result.code,
                    map: result.map || null,
                };
            }
            catch (error) {
                const err = error;
                if (viteServer) {
                    viteServer.ws.send({
                        type: "error",
                        err: {
                            message: err.message,
                            stack: err.stack || "",
                            plugin: "vite-plugin-fynix",
                            id,
                        },
                    });
                }
                console.error(`\x1b[32m[vite-plugin-fynix]\x1b[0m \x1b[31mFailed to transform ${id}:\x1b[0m`);
                console.error(err.message);
                return null;
            }
        },
        handleHotUpdate(ctx) {
            const { file, server } = ctx;
            const normalizedFile = normalizePath(file);
            const shouldReload = include.some((ext) => normalizedFile.endsWith(ext));
            if (shouldReload) {
                console.log(`\x1b[32m[vite-plugin-fynix]\x1b[0m HMR: full-reload triggered by ${normalizedFile}`);
                server.ws.send({
                    type: "full-reload",
                    path: "*",
                });
                return [];
            }
            return undefined;
        },
        config() {
            return {
                esbuild: false,
                optimizeDeps: {
                    include: ["fynixui"],
                    esbuildOptions: {
                        jsx: "transform",
                        jsxFactory,
                        jsxFragment,
                    },
                },
                resolve: {
                    extensions: [".fnx", ".ts", ".tsx", ".js", ".jsx", ".json"],
                },
            };
        },
        buildStart() {
            console.log(`\x1b[32m[vite-plugin-fynix]\x1b[0m Initialized with JSX factory: ${jsxFactory}`);
        },
    };
}
export { fynixPlugin };
